### PR TITLE
Port Filter/Target changes

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This is a pretty liberal port of https://github.com/firebase/firebase-android-sdk/pull/3237/files that also brings in a bunch of Indexing functionality from the Android SDK (as the PR touches a bunch of this functionality).

The PR also cleans up an existing naming issue of `isDocumentTarget` and renames it to follow the pattern className+methodName. In the tests, the PR simplifies the `bound` method.

### CODE REVIEW

**Comment by ehsannas** [comment]
Sorry I've fallen behind on this. I'll take a look soon.

**Review by ehsannas — APPROVED** [review]

File: packages/firestore/src/core/target.ts [hunk 1]
```diff
@@ -187,14 +198,261 @@ export function targetEquals(left: Target, right: Target): boolean {
...
+          // Encode equality prefix, which is encoded in the index value before
+          // the inequality (e.g. `a == 'a' && b != 'b'` is encoded to
+          // `value != 'ab'`).
+          values.push(fieldFilter.value);
+          break;
+        case Operator.NOT_IN:
+        case Operator.NOT_EQUAL:
+          // NotIn/NotEqual is always a suffix
+          values.push(fieldFilter.value);
+          return values;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/core/target.ts:275]: I had to think about this a bit when I ported (and also recently added a unit test for it - https://github.com/firebase/firebase-android-sdk/pull/3351/files#diff-216e85d9cb7dc6a9ca6b6bf1ae2d9e7751d20a13802e9a7c97a8f517b0bff67cR128). The index encoding order is not specified by the query but rather the index. The order that the values is encoded in is specific to the index.

[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/core/target.ts:275]: I will try to improve the comment.

[the remark this answers is not in the document]
Comment (ehsannas) [reply packages/firestore/src/core/target.ts:275]: 👍  Thanks

File: packages/firestore/src/core/target.ts [hunk 2]
```diff
@@ -187,14 +198,261 @@ export function targetEquals(left: Target, right: Target): boolean {
...
+      }
+    }
+  }
+
+  return null;
+}
+
+/**
+ * Returns a lower bound of field values that can be used as a starting point to
+ * scan the index defined by `fieldIndex`}`. Returns `null` if no lower bound
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/core/target.ts]: Fixed

File: packages/firestore/src/model/values.ts [hunk 3]
```diff
@@ -602,3 +606,87 @@ export function deepClone(source: Value): Value {
...
+  } else {
+    return fail('Invalid value type: ' + JSON.stringify(value));
+  }
+}
+
+export function valuesMax(
+  left: Value | undefined,
+  right: Value | undefined
+): Value | undefined {
+  if (left === undefined && right === undefined) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/values.ts]: Had to think about this for a bit but you are right. Fixed.

**Pull request merged**